File: solarplane.m
# Parameters of the Airplane

#Mass
                  # wing  fuse/tail  solarcell  radio
com.m.structure =   0.1  +  0.05    +  0.07    + 0.09;

#Geometory
com.b = 2.4;
com.S = 0.4;
com.AR = com.b**2/com.S;
com.A_fuse = 0.05*0.05;

#Aerodynamic

com.Cd0_wing = 0.02;
com.Cdp_fuse = 0.1;
com.CL_min = 0.3;
com.CL_max = 1.2;

function CD = CD(CL,com)
         #  CL_2Dwing                       # 3D property
	CD = com.Cd0_wing + 0.015*(CL-0.25)**2 + CL**2/(0.9*pi*com.AR) + com.Cdp_fuse*com.A_fuse/com.S;
endfunction

function [L D] = LD(Vc,CL,com)
    q = 0.5*com.rho*Vc**2;
    L = q*CL*com.S;
    D = q*CD(CL,com)*com.S;
endfunction

File: modules/SolarPlane.jl
module SolarPlaneModule
export SolarPlane, LD, CD

struct SolarPlane
m_wing::Float64
m_empn::Float64
m_mech::Float64
b::Float64
S::Float64
AR::Float64
A_fuse::Float64
Cd0_wing::Float64
Cdp_fuse::Float64
CL_min::Float64
CL_max::Float64
end

function SolarPlane(data::Dict)
  required_fields = ["m_wing",
                     "m_empn", # fuse+tail
                     "m_mech",
                     "b", "S", "A_fuse",
                     "Cd0_wing", "Cdp_fuse", "CL_min", "CL_max"]
  for field in required_fields
    if !haskey(data, field)
      error("Missing required field '$field' in SolarPlane data")
    end
  end

  AR = data["b"]^2 / data["S"]
  SolarPlane(
             Float64(data["m_wing"]),
             Float64(data["m_empn"]),
             Float64(data["m_mech"]),
             Float64(data["b"]),
             Float64(data["S"]),
             AR,
             Float64(data["A_fuse"]),
             Float64(data["Cd0_wing"]),
             Float64(data["Cdp_fuse"]),
             Float64(data["CL_min"]),
             Float64(data["CL_max"])
             )
end

function CD(CL, solarplane::SolarPlane)
  #                            2Dwing                  CDi
  return solarplane.Cd0_wing + 0.015 * (CL - 0.25)^2 + CL^2 / (0.9 * pi * solarplane.AR) + solarplane.Cdp_fuse * solarplane.A_fuse / solarplane.S
end

function LD(Vc, CL, solarplane::SolarPlane, common)
  q = 0.5 * common.rho * Vc^2
  L = q * CL * solarplane.S
  D = q * CD(CL, solarplane) * solarplane.S
  return L, D
end

end # module
